host/bcm283x/gpio.go

```diff
@@ -395,58 +388,54 @@ func (p *Pin) PWM(duty gpio.Duty, period time.Duration) error {
 	if pwmMemory == nil || clockMemory == nil {
 		return p.wrap(errors.New("bcm283x-dma not initialized; try again as root?"))
 	}
+	p.usingClock = true
+	base_freq := uint64(25 * 1000 * 1000) // 25MHz
 
-	// Convert period to frequency. This is lossy.
-	hz := uint64(time.Second / period)
+	if useDMA {
+		dma_freq := uint64(200 * 1000) // 200KHz
+		div := uint32(base_freq / dma_freq)
+		// Total cycles in the period
+		rng := dma_freq * uint64(period) / uint64(time.Second)
+		// Pulse width cycles
+		dat := uint32(rng * uint64(duty) / uint64(gpio.DutyMax))
 
-	if clkID != -1 {
-		// Using a clock pin.
-		var clk *clock
-		switch clkID {
-		case 0:
-			clk = &clockMemory.gp0
-		case 2:
-			clk = &clockMemory.gp2
-		}
-		actual, waits, err := clk.set(hz, 1)
-		if err != nil {
+		var err error
+		// TODO(simokawa): Reuse DMA buffer if possible.
```

Comment:
The thing I'm the most worried is to leave the host on a broken state. One thing that would encourage me is to have bcm283xsmoketest that turns on PWM on and off 20 times. If that works I'll feel better about it.

Reply:
I updated smoketest and run it 20 times. Looks good so far.